### PR TITLE
Extend tlsfuzzer coverage

### PR DESCRIPTION
**Pull request by Jakuje**
#### Description

This mostly extends the test coverage for TLS 1.2 and ECDSA and EdDSA keys. I did not touch the RSA-PSS coverage, as it needs to use the special RSA-PSS keys and certificates to work correctly and pkcs11-provider can not use them. This will likely be handled in some follow-up PR.

#### Checklist

- [X] Test suite updated with functionality tests
- [X] Test suite updated with negative tests

#### Reviewer's checklist:

- [x] Any issues marked for closing are addressed
- [x] There is a test suite reasonably covering new functionality or modifications
- [x] This feature/change has adequate documentation added
- [x] Code conform to coding style that today cannot yet be enforced via the check style test
- [x] Commits have short titles and sensible commit messages
- [x] Coverity Scan has run if needed (code PR) and no new defects were found

### CODE REVIEW

**Review by simo5 — DISMISSED** [review]
LGTM!

Covscan not needed.

File: tests/cert.json.ecdsa.in [hunk 1]
```diff
@@ -0,0 +1,21 @@
+,
+    {"server_command": [@CHECKER@"openssl", "s_server", @PROPQ@"-www", "-port", "@PORT@", "-key", "@PRIURI@", "-cert", "@CRTURI@"],
+     "comment": "Run test with @PRIURI@ without certificate verify",
+     "environment": {"PYTHONPATH" : "."},
+     "server_hostname": "localhost",
+     "server_port": @PORT@,
+     "tests" : [
+       {"name" : "test-tls13-conversation.py",
+        "arguments" : ["-p", "@PORT@"]},
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply tests/cert.json.ecdsa.in:9]: Thanks! This is certainly helpful to save some writing!

File: tests/cert.json.ecdsa.in [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+       {"name" : "test-conversation.py",
+        "arguments" : ["-p", "@PORT@", "-d"]},
+       {"name" : "test-signature-algorithms.py",
+        "arguments" : [
+          "-p", "@PORT@", "-n", "0", "--ecdsa", "-g", "secp384r1", "-x", "sanity",
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply tests/cert.json.ecdsa.in:14]: But the test with `test-signature-algorithms.py  -n 0 --ecdsa` says it tests ECDSA+SHA1, which wont work in Fedora/RHEL with default crypto policies, isn't it? So then we should probably drop both executions of this script?

**Comment by Jakuje** [comment]
> there are multiple scripts missing compared to instructions from https://github.com/openssl/openssl/pull/25724

For RSA, I added `test-dhe-rsa-key-exchange-signatures.py`, but again, the SHA1 signatures are failing per default c-p so I can add that, but will have to skip these:
```
test-dhe-rsa-key-exchange-signatures.py:stdout: 'TLS_DHE_RSA_WITH_AES_128_CBC_SHA256 sha1 signature'
test-dhe-rsa-key-exchange-signatures.py:stdout: 'TLS_DHE_RSA_WITH_AES_128_CBC_SHA sha1 signature'
test-dhe-rsa-key-exchange-signatures.py:stdout: 'TLS_DHE_RSA_WITH_AES_256_CBC_SHA256 sha1 signature'
test-dhe-rsa-key-exchange-signatures.py:stdout: 'TLS_DHE_RSA_WITH_AES_256_CBC_SHA sha1 signature'
```
The test `test-signature-algorithms.py` again talks about SHA-1 signatures. Is there any point in keeping it or ok to skip altogether?

File: tests/cert.json.ecdsa.in (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomato42) [reply tests/cert.json.ecdsa.in]: no, that script also verifies that all the different hashes can be used, you can expect failures in the sha-1 cases (adding a `"comment": "Crypto-Policies disable SHA-1"` would be nice too)

File: tests/cert.json.ecdsa.in [hunk 3]
```diff
@@ -0,0 +1,21 @@
+,
+    {"server_command": [@CHECKER@"openssl", "s_server", @PROPQ@"-www", "-port", "@PORT@", "-key", "@PRIURI@", "-cert", "@CRTURI@"],
+     "comment": "Run test with @PRIURI@ without certificate verify",
+     "environment": {"PYTHONPATH" : "."},
+     "server_hostname": "localhost",
+     "server_port": @PORT@,
+     "tests" : [
```
[the remark this answers is not in the document]
Comment (tomato42) [reply tests/cert.json.ecdsa.in:8]: yes, add them with `-x 'connect with ecdsa_brainpoolP256r1tls13_sha256 only' -X handshake_failure`

**Comment by tomato42** [comment]
> The test test-signature-algorithms.py again talks about SHA-1 signatures. Is there any point in keeping it or ok to skip altogether?

if there are other cases that do pass, yes, modifying it with `-x ... -X ...` to expect failures is completely valid way to use those scripts

File: tests/cert.json.ecdsa.in [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+     "server_port": @PORT@,
+     "common_arguments": ["-p", "@PORT@"],
+     "tests" : [
+       {"name" : "test-tls13-conversation.py"},
+       {"name" : "test-conversation.py",
+        "arguments" : ["-d"]},
+       {"name" : "test-ecdsa-sig-flexibility.py",
+        "arguments" : [
+          "-n", "0",
+          "-e", "connect with ecdsa_brainpoolP256r1tls13_sha256 only",
```
Comment (tomato42) [inline tests/cert.json.ecdsa.in:15]: I think it would be better to use `-x` here

File: tests/cert.json.ecdsa.in [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+          "-e", "connect with ecdsa_brainpoolP384r1tls13_sha384 only",
+          "-e", "connect with ecdsa_brainpoolP512r1tls13_sha512 only",
+          "-e", "connect with sha1+ecdsa only"
+        ]
+       },
+       {"name" : "test-signature-algorithms.py",
```
Comment (tomato42) [inline tests/cert.json.ecdsa.in:21]: `test-signature-algorithms.py` without `-g` still missing?

File: tests/cert.json.rsa.in [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        "arguments" : ["-d"]},
+       {"name" : "test-dhe-rsa-key-exchange-signatures.py",
+        "arguments" : [
+          "-n", "0",
+          "-x", "TLS_DHE_RSA_WITH_3DES_EDE_CBC_SHA sha1 signature", "-X", "handshake_failure",
+          "-x", "TLS_DHE_RSA_WITH_3DES_EDE_CBC_SHA sha224 signature", "-X", "handshake_failure",
+          "-x", "TLS_DHE_RSA_WITH_3DES_EDE_CBC_SHA sha256 signature", "-X", "handshake_failure",
+          "-x", "TLS_DHE_RSA_WITH_3DES_EDE_CBC_SHA sha384 signature", "-X", "handshake_failure",
+          "-x", "TLS_DHE_RSA_WITH_3DES_EDE_CBC_SHA sha512 signature", "-X", "handshake_failure",
+          "-e", "TLS_DHE_RSA_WITH_AES_128_CBC_SHA256 sha1 signature",
```
Comment (tomato42) [inline tests/cert.json.rsa.in:20]: why `-e` and not `-x`?

Comment (Jakuje) [reply tests/cert.json.rsa.in:20]: My though was that by excluding the tests, we say "we do not care for the results", do not depend on particular openssl configuration and we could make it working on other distros/os than default fedora with default crypto policies that might allow SHA1 or brainpool. Right now CI runs also on Debian, Ubuntu, CentOS and MacOS.

File: tests/cert.json.ecdsa.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jakuje) [reply tests/cert.json.ecdsa.in:8]: I skipped the brainpool, as expecting failure on these does not seem to work:
```
test-ecdsa-sig-flexibility.py:stdout:connect with ecdsa_brainpoolP256r1tls13_sha256 only ...
test-ecdsa-sig-flexibility.py:stdout:Error encountered while processing node ExpectAlert(level=2, description=40) (child: <tlsfuzzer.expect.ExpectClose object at 0x7fc6619c96a0>) with last message being: <tlslite.messages.Message object at 0x7fc661c82950>
test-ecdsa-sig-flexibility.py:stdout:Error while processing
server:stderr:80629355D17F0000:error:0A000126:SSL routines::unexpected eof while reading:ssl/record/rec_layer_s3.c:689:
test-ecdsa-sig-flexibility.py:stdout:Traceback (most recent call last):
test-ecdsa-sig-flexibility.py:stdout:  File "/home/jjelen/devel/pkcs11-provider/tlsfuzzer/scripts/test-ecdsa-sig-flexibility.py", line 287, in main
test-ecdsa-sig-flexibility.py:stdout:    runner.run()
test-ecdsa-sig-flexibility.py:stdout:    ~~~~~~~~~~^^
test-ecdsa-sig-flexibility.py:stdout:  File "/home/jjelen/devel/pkcs11-provider/tlsfuzzer/tlsfuzzer/runner.py", line 235, in run
test-ecdsa-sig-flexibility.py:stdout:    raise AssertionError("Unexpected message from peer: " +
test-ecdsa-sig-flexibility.py:stdout:    ...<4 lines>...
test-ecdsa-sig-flexibility.py:stdout:                                        RecordHeader2)))
test-ecdsa-sig-flexibility.py:stdout:AssertionError: Unexpected message from peer: Handshake(server_hello)
test-ecdsa-sig-flexibility.py:stdout:
test-ecdsa-sig-flexibility.py:stdout:Expected error message: handshake_failure
```
Not sure if this is an issue of the pkcs11-provider, openssl or expected behavior.

File: tests/cert.json.ecdsa.in (hunk 5 above)
Comment (Jakuje) [reply tests/cert.json.ecdsa.in:21]: Added with skipping the failing tests.

File: tests/cert.json.rsa.in (hunk 6 above)
Comment (Jakuje) [reply tests/cert.json.rsa.in]: Sounds like here expecting failure works as expected, so resolving this one.

File: tests/cert.json.ecdsa.in (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jakuje) [reply tests/cert.json.ecdsa.in]: Added.

File: tests/cert.json.ecdsa.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomato42) [reply tests/cert.json.ecdsa.in:8]: if we don't care about brainpool support in PKCS#11, then yes, skipping them (with `-e` is probably better)

**Comment by tomato42** [comment]
> > also, I think it would be nice to check that PKCS#11 can be used to delegate verification of signatures too...?
> 
> AFAIK this is implicitly tested by adding the propquery to the openssl cli in the second invocation of the tests:
> 
> https://github.com/latchset/pkcs11-provider/blob/6be8f7aedc119d5ceb7b9e4db5c4df8b73ee46af/tests/ttlsfuzzer#L81C1-L81C6
> 
> This should force the OpenSSL to import even public keys to the pkcs11 provider and do the operations on them. This already caught us some architectural issues in the past that have been fixed.

ok, but isn't the point of those tests to ensure end-to-end support for that, in the TLS context specifically?

File: tests/cert.json.rsa.in (hunk 6 above)
Comment (tomato42) [reply tests/cert.json.rsa.in]: no, I think the idea is to verify that SHA-1 actually _doesn't_ work, for that we need `-x`

File: tests/cert.json.ecdsa.in [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        "arguments" : [
+          "-n", "0", "--ecdsa",
+          "-x", "duplicated 206 non-rsa schemes", "-X", "handshake_failure",
+          "-x", "duplicated 2346 non-rsa schemes", "-X", "handshake_failure",
+          "-x", "duplicated 8123 non-rsa schemes", "-X", "handshake_failure",
+          "-x", "duplicated 23745 non-rsa schemes", "-X", "handshake_failure",
+          "-x", "duplicated 32748 non-rsa schemes", "-X", "handshake_failure",
+          "-x", "explicit SHA-256+RSA or ECDSA", "-X", "handshake_failure",
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply tests/cert.json.ecdsa.in:30]: @tomato42 This looks like the OpenSSL is looking at the signature in the certificate and given that we have RSA CA and EC key in the TLS server, it contains the `Signature Algorithm: sha256WithRSAEncryption`, which makes this connection to fail.

Tried the same test with ECDSA CA key and these work ok, but other bunch of RSA tests start failing when we have EC CA and RSA host key:
```
test-tls13-rsa-signatures.py:stdout:FAILED:
test-tls13-rsa-signatures.py:stdout:    'sanity'
test-tls13-rsa-signatures.py:stdout:    'sanity'
test-tls13-rsa-signatures.py:stdout:    'tls13 signature rsa_pss_rsae_sha256'
test-tls13-rsa-signatures.py:stdout:    'tls13 signature rsa_pss_rsae_sha384'
test-tls13-rsa-signatures.py:stdout:    'tls13 signature rsa_pss_rsae_sha512'

test-tls13-signature-algorithms.py:stdout:FAILED:
test-tls13-signature-algorithms.py:stdout:      'sanity'
test-tls13-signature-algorithms.py:stdout:      'sanity'
test-tls13-signature-algorithms.py:stdout:      'tolerance 32715 methods with sig_alg_cert'
test-tls13-signature-algorithms.py:stdout:      'tolerance legacy RSA PKCS#1.5'
test-tls13-signature-algorithms.py:stdout:      'tolerance unallocated 0x0A01 (10+RSA) method'
test-tls13-signature-algorithms.py:stdout:      'unique and well-known sig_algs, rsa algorithms last'

test-sig-algs.py:stdout:FAILED:
test-sig-algs.py:stdout:        'extra sigalgs'
test-sig-algs.py:stdout:        'MD5 first'
test-sig-algs.py:stdout:        'MD5 first, no SHA-1'
test-sig-algs.py:stdout:        'no SHA-1'
test-sig-algs.py:stdout:        'rsa_pss_rsae_sha256 only'
test-sig-algs.py:stdout:        'rsa_pss_rsae_sha384 only'
test-sig-algs.py:stdout:        'rsa_pss_rsae_sha512 only'
test-sig-algs.py:stdout:        'sanity'
test-sig-algs.py:stdout:        'sanity'

```
And some of the other tests do not seem to work at all.


Is the tlsfuzzer usage limited to use with self-signed certificates or certificates signed with the same key type? Or is there some bug in tlsfuzzer or openssl to cover the combination?

I do not think we want to change the whole testsuite to have self-signed certificates or to have different CA certificate chains just for tlsfuzzer @simo5  so I think this state is the best we can get now.

[the remark this answers is not in the document]
Comment (tomato42) [reply tests/cert.json.ecdsa.in:30]: there is no explicit policy for tlsfuzzer that the certificates have to be self-signed or anything like that

if openssl ignores `signature_algorithms_cert` and uses `signatures_algorithms` for selection of the server certificate, and we want to have heterogenous keys in the certificate chain, then the only option is to modify the test script to send both sha256+rsa and sha256+ecdsa, but then expect only one of the (depending on the `--ecdsa` option) in the ServerKeyExchange

all of that is already supported, just the test script needs to be updated

[the remark this answers is not in the document]
Comment (simo5) [reply tests/cert.json.ecdsa.in:30]: Ok I think we should merge as is then, and then will have a folllwup PR once the tlsfuzzer script can be updated to handle these cases.

**Review by simo5 — APPROVED** [review]
LGTM

File: tests/cert.json.ecdsa.in (hunk 7 above)
[the remark this answers is not in the document]
Comment (Jakuje) [reply tests/cert.json.ecdsa.in:30]: Fine with me! Not sure if there is some followup related to the other comment from Alicja, but I think this should be good enough for now.

[the remark this answers is not in the document]
Comment (tomato42) [reply tests/cert.json.ecdsa.in:30]: I'd say there is... to have it actually tested we need to either have homogenous CA and server keys for the RSA and ECDSA cases, or update the tlsfuzzer scripts

I suggest the former.

but I don't think it's done unless either of those options is followed

[the remark this answers is not in the document]
Comment (Jakuje) [reply tests/cert.json.ecdsa.in:30]: Filled #490 with the follow-up work. @tomato42 please take a look when time permits.

**Pull request merged**